osu.Game/Rulesets/RulesetStore.cs

```diff
@@ -113,64 +107,73 @@ private void addMissingRulesets()
             }
         }
 
-        private static void loadFromAppDomain()
+        /// <summary>
+        /// Loads the rulesets that are in the current appdomain an in the current directory.
+        /// </summary>
+        /// <returns>The rulesets that were loaded.</returns>
+        private static Dictionary<Assembly, Type> loadRulesets()
+        {
+            var rulesets = new Dictionary<Assembly, Type>();
+
+            foreach (var rulesetAssembly in getRulesetAssemblies())
+            {
+                try
+                {
+                    rulesets[rulesetAssembly] = rulesetAssembly.GetTypes().First(t => t.IsPublic && t.IsSubclassOf(typeof(Ruleset)));
+                }
+                catch (Exception e)
+                {
+                    Logger.Error(e, $"Failed to add ruleset {rulesetAssembly}");
+                }
+            }
+
+            return rulesets;
+        }
+
+        /// <summary>
+        /// Scans the current appdomain and current directory for ruleset assemblies.
+        /// Rulesets that were found in the current directory are automaticly loaded.
+        /// </summary>
+        /// <returns>The ruleset assemblies that were found in the current appdomain or in the current directory.</returns>
+        private static IEnumerable<Assembly> getRulesetAssemblies()
         {
+            var rulesetAssemblies = new HashSet<Assembly>();
+
+            // load from appdomain
             foreach (var ruleset in AppDomain.CurrentDomain.GetAssemblies())
             {
                 string rulesetName = ruleset.GetName().Name;
 
                 if (!rulesetName.StartsWith(ruleset_library_prefix, StringComparison.InvariantCultureIgnoreCase) || ruleset.GetName().Name.Contains("Tests"))
                     continue;
 
-                addRuleset(ruleset);
+                rulesetAssemblies.Add(ruleset);
             }
-        }
 
-        private static void loadFromDisk()
-        {
+            // load from current directory
             try
             {
                 string[] files = Directory.GetFiles(Environment.CurrentDirectory, $"{ruleset_library_prefix}.*.dll");
 
                 foreach (string file in files.Where(f => !Path.GetFileName(f).Contains("Tests")))
-                    loadRulesetFromFile(file);
+                {
+                    try
+                    {
+                        rulesetAssemblies.Add(Assembly.LoadFrom(file));
```

Comment:
So I assume the `Lazy` was added because `OsuGameActivity` doesn't get the chance to overwrite `Environment.CurrentDirectory` before `RulesetStore` starts looking. Leaving aside the hackiness of that (is that execution order guaranteed?), can xamarin even dynamically load in assemblies from DLLs?

Reply:
>can xamarin even dynamically load in assemblies from DLLs?
yes

Reply:
> So I assume the Lazy was added because OsuGameActivity doesn't get the chance to overwrite Environment.CurrentDirectory before RulesetStore starts looking.

That is correct. Enforcing the direct path ensures that the working directory will point to correct directory.

> is that execution order guaranteed?

The line https://github.com/ppy/osu/pull/6510/files#diff-1afa33156b358a1e34958fc4fbdadf87R23 gets called before the framework loads. (The framework gets loaded in the base.OnCreate)
Therefore yes.